### PR TITLE
Added rd_kafka_list_groups API to get consumer group subscription metadata

### PR DESCRIPTION
**Pull request by simnotes**
I added new GetConsumerGroupsMetadata function to consumer, producer and adminClient API, which calls rd_kafka_list_groups from librdkafka to retrieve current consumer group status and consumer group subscriptions from broker. Via the new API it is possible to query subscription-status like in "kafka-consumer-groups.sh"

Also added integration and dry-tests to everything.

(somehow this is a copy form pr #331 )

### CODE REVIEW

**Comment by simnotes** [comment]
@edenhill I published the included changes prior to the 1.0.0 release and you told me, that those changes might be merged after the release (which is understandable). So now I merged the 1.0.0 release with my changes, fixed the merge issues, ran the tests again and everything seems to be ok -> Do you mind, having a look again?

File: kafka/integration_test.go [hunk 1]
```diff
@@ -1559,3 +1559,198 @@ func TestAdminGetMetadata(t *testing.T) {
...
+	t.Logf("Creating consumer and subscribe to test-topic")
+	c, err := NewConsumer(&consumerConf)
+	if err != nil {
+		t.Fatalf("NewConsumer: %v", err)
+	}
+	defer c.Close()
+
+	c.Subscribe(testconf.Topic, nil)
+
+	// FIXME: wait for topics to become available in metadata
```
Comment (edenhill) [inline kafka/integration_test.go]: Dont think this is needed since the consumer is created after the producer is done producing to the topics.

File: kafka/integration_test.go [hunk 2]
```diff
@@ -1559,3 +1559,198 @@ func TestAdminGetMetadata(t *testing.T) {
...
+	if err != nil {
+		t.Errorf("Failed to get consumer group metadata for Group %s. Error %s\n", testconf.GroupID, err)
+	}
+	t.Logf("Consumer group meta data for group %s was: %v\n", testconf.GroupID, cgMetadata)
+
+	// Try to find test-topic name in metadata
+	found := false
+	for _, topic := range cgMetadata[0].Members[0].MemberMetadata.Topics {
+		if topic == testconf.Topic {
+			found = true
```
Comment (edenhill) [inline kafka/integration_test.go]: should it verify that the consumer is seen too?

File: kafka/integration_test.go [hunk 3]
```diff
@@ -1559,3 +1559,198 @@ func TestAdminGetMetadata(t *testing.T) {
...
+		"acks":                1}
+
+	producerConf.updateFromTestconf()
+
+	p, err := NewProducer(&producerConf)
+	if err != nil {
+		panic(err)
+	}
+
+	// FIXME: wait for topics to become available in metadata
```
Comment (edenhill) [inline kafka/integration_test.go]: dito

File: kafka/metadata.go [hunk 4]
```diff
@@ -68,12 +77,35 @@ type TopicMetadata struct {
 
 // Metadata contains broker and topic metadata for all (matching) topics
 type Metadata struct {
-	Brokers []BrokerMetadata
-	Topics  map[string]TopicMetadata
-
+	Brokers           []BrokerMetadata
```
Comment (edenhill) [inline kafka/metadata.go]: unrelated whitespace change, remove.

File: kafka/adminapi.go [hunk 5]
```diff
@@ -859,6 +859,14 @@ func (a *AdminClient) GetMetadata(topic *string, allTopics bool, timeoutMs int)
 	return getMetadata(a, topic, allTopics, timeoutMs)
 }
 
+// GetConsumerGroupMetadata queries broker for consumer groups metadata and all its members
+// like which consumer groups and which corresponding members consume which topics
+// if consumerGroup is nil, metadata for all consumer groups will be returned
+// otherwise only metadata for that specific consumer group will be returned
+func (a *AdminClient) GetConsumerGroupMetadata(consumerGroup *string, timeoutMs int) ([]ConsumerGroupMetadata, error) {
```
Comment (edenhill) [inline kafka/adminapi.go]: I think this should probably be called DescribeConsumerGroup() to be somewhat consistent with https://cwiki.apache.org/confluence/display/KAFKA/KIP-222+-+Add+Consumer+Group+operations+to+Admin+API

File: kafka/producer.go [hunk 6]
```diff
@@ -572,6 +572,14 @@ func (p *Producer) QueryWatermarkOffsets(topic string, partition int32, timeoutM
 	return queryWatermarkOffsets(p, topic, partition, timeoutMs)
 }
 
+// GetConsumerGroupMetadata queries broker for consumer groups metadata and all its members
+// like which consumer groups and which corresponding members consume which topics
+// if consumerGroup is nil, metadata for all consumer groups will be returned
+// otherwise only metadata for that specific consumer group will be returned
+func (p *Producer) GetConsumerGroupMetadata(consumerGroup *string, timeoutMs int) ([]ConsumerGroupMetadata, error) {
```
Comment (edenhill) [inline kafka/producer.go]: We don't need to provide this method on Producer and Consumer since an AdminClient can be instantiated from an existing Producer/Consumer without much cost.

**Comment by nsagnett** [comment]
Hi !

Any news on this topic ? It should be a great feature !

**Comment by milindl** [comment]
Closing this now that rd_kafka_list_groups is deprecated and replaced with rd_kafka_ListConsumerGroups and rd_kafka_DescribeConsumerGroups. 

Both are available in the go client via #923.
